### PR TITLE
IAM-943

### PR DESCRIPTION
**Pull request by shipperizer**
IAM-943: hydra internal networking

- **feat: introduce internal ingress**
- **fix: use internal ingress if set, otherwise stick with k8s networking**
- **fix: patch if secret not found**

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -528,19 +636,14 @@ def _update_hydra_endpoints_relation_data(self, event: RelationEvent) -> None:
         logger.info("Sending endpoints info")
 
         admin_endpoint = (
-            self._admin_url
+            self._internal_url
             or f"http://{self.app.name}.{self.model.name}.svc.cluster.local:{HYDRA_ADMIN_PORT}"
         )
```
Comment (nsklikas) [inline src/charm.py:641]: The `/admin` part is not needed?

File: src/charm.py [hunk 2]
```diff
@@ -1009,6 +1117,32 @@ def _get_tracing_endpoint_info(self) -> str:
...
+
+        Args:
+            event: a :class:`HookEvent` to signal a change we may need to respond to.
+        """
+        if not self.unit.is_leader():
+            return
+
+        # If it's a RelationJoinedEvent, set it in the ingress object
+        if isinstance(event, RelationJoinedEvent):
+            self.internal_ingress._relation = event.relation
```
Comment (nsklikas) [inline src/charm.py:1136]: Why is this needed?

File: src/charm.py [hunk 3]
```diff
@@ -228,6 +241,14 @@ def __init__(self, *args: Any) -> None:
             self.loki_consumer.on.promtail_digest_error,
             self._promtail_error,
         )
+        self.framework.observe(
+            self.on[INTERNAL_INGRESS_RELATION_NAME].relation_joined, self._configure_ingress
+        )
+        self.framework.observe(
+            self.on[INTERNAL_INGRESS_RELATION_NAME].relation_changed, self._configure_ingress
+        )
```
Comment (nsklikas) [inline src/charm.py:249]: Shouldn't we handle relation revoked as well?

**Review by nsklikas — COMMENTED** [review]
lgtm overall, some minor questions

File: src/charm.py (hunk 2 above)
Comment (shipperizer) [reply src/charm.py:1136]: not sure to be fair, implementation is matching grafana's one

happy to remove and try to see if it works

File: src/charm.py (hunk 1 above)
Comment (shipperizer) [reply src/charm.py:641]: nope, the /admin is added by client sdks

if we do it then it is like adding and extra one when calling the admin API so it would end up being `/admin/admin/clients`

File: src/charm.py (hunk 2 above)
Comment (shipperizer) [reply src/charm.py:1139]: chatted and all hell breaks loose if removed

**Review by nsklikas — APPROVED** [review]
